### DISCUSSION TITLE
Ignore certain files for changes on_serve or stop next rebuild until the current finishes

### DISCUSSION THREAD

**Discussion by mondeja** [mkdocs/mkdocs#2544] [Q&A] [answered]
I'm developing a plugin that, essentially, do: from a file `foo.md` inside the root of the documentation directory, creates another file `foo.ext`, if it doesn't exist previously. So the markdown files of the documentation are linked to other files.

The problem is: if I edit `foo.md` or `foo.ext` after executing `mkdocs serve` it triggers an infinite loop of rebuilds. I've tried a lot of possible patchs with `mkdocs==1.2.0`, `mkdocs==1.2.1` and `mkdocs==1.2.2`, but nothing worked.

It seems strange to me that changes are detected before the previous build is complete. Makes this any sense or is an implementation lack? As the maintainer of livereloading, could you provide some explanation about this behaviour @oprypin? Additionally, is possible to ignore certain files from being observed inside a observed directory? I suspect that #2529 could help, but determine `*.ext` files as hidden ones still seems like a hack. If so, I think that should be documented. Thanks!

---

### Replicable example

```
git clone https://github.com/mondeja/mkdocs-mdpo-plugin.git
cd mkdocs-mdpo-plugin
python3 -m virtualenv venv
, venv/bin/activate
python3 -m pip install -e .[dev]
python3 -m mkdocs serve
```

**Comment by oprypin** [comment, marked as answer]
> I'm developing a plugin that, essentially, do: from a file `foo.md` inside the root of the documentation directory, creates another file `foo.ext`

It is indeed a problem if you do that. It's not a good approach, and I have [criticized it before and mentioned an alternative](https://github.com/mkdocs/mkdocs/issues/2285#issuecomment-815978652):

> the exact described use case (writing files into docs dir during the build of that docs dir) is not a good approach and, I'd say, should not be catered towards.
> There will be other issues, such as the file accidentally overwriting real files, or lingering there after an interruption and accidentally being committed, and probably more.
> MkDocs' behavior isn't even wrong there.
>
> Instead one can write to a separate temporary directory and add *that* file into the Files structure.
> https://github.com/oprypin/mkdocs-gen-files/blob/08d6e91924af3a67031f30d925de12e893c5804b/mkdocs_gen_files/editor.py#L42-L47.

https://github.com/oprypin/mkdocs-gen-files/blob/08d6e91924af3a67031f30d925de12e893c5804b/mkdocs_gen_files/plugin.py#L27-L29
https://github.com/oprypin/mkdocs-gen-files/blob/08d6e91924af3a67031f30d925de12e893c5804b/mkdocs_gen_files/editor.py#L114-L115

To reiterate: don't programmatically write into `docs_dir`, instead write to a temporary location and add a `File` object referencing that location.

You can check out the code linked above -- do that instead of this code (with a mostly untrue comment ;)
https://github.com/mondeja/mkdocs-mdpo-plugin/blob/5dc80c340b436011063b21348679a4c8dd646763/mkdocs_mdpo_plugin/plugin.py#L311-L333

---

And then the described issue about livereload becomes invalid.

**Reply by mondeja** [reply]
Thank you for the detailed response and the great work overall, really appreciated :pray:

**Comment by KonstantinNosov** [comment]
May be it will be useful for someone.
I ran into exactly same issue due to mkdocs materal blog plugin which manipulates with files on FS instead of handling abstraction registered with on_files. I need to generated blog posts so had to deal with it. 

Below is the code that removes mkdocs watch for docs_dir directory (or any other)  and starts watching the excluded directory  maintaining mkdocs individual  watches for each file in it except those in subdir to be ignored:

```python

def mkdocs_watch_ignore_path( server: LiveReloadServer, config: MkDocsConfig, ignore_dir, watched_dir=None):
    if watched_dir is None:
        watched_dir=config.docs_dir

    try:
        # Unwatch the directory watched by mkdocs.
        server.unwatch(watched_dir)
    except KeyError:
        # watched_dir is not watched
        pass

    def callback(event: FileSystemEvent):
        # Ignore events for directories as they do not affect mkdocs livereload
        if event.is_directory:
            return

        # Ignore events for files within `ignore_dir`
        if Path(event.src_path).is_relative_to(ignore_dir):
            return

        # Unwatch deleted files
        if event.event_type in  [watchdog.events.EVENT_TYPE_DELETED,watchdog.events.EVENT_TYPE_MOVED]:
            try:
                server.unwatch(event.src_path)
            except KeyError:
                # src_path is not watched
                pass

        # Watch created files
        if event.event_type in  [watchdog.events.EVENT_TYPE_CREATED]:
            server.watch(event.src_path)
        if event.event_type in  [watchdog.events.EVENT_TYPE_MOVED]:
            server.watch(event.dest_path)

        # Touch mkdocs config file which is always watched to trigger rebuild and handle created/deleted files
        os.utime(config.config_file_path)
        return

    # Call mkdocs watch for files within in `watched_dir` tree except those in `ignore_dir`
    for path in glob.glob(os.path.join(watched_dir, "**/*"), recursive=True):
        if  os.path.isfile(path) and not Path(path).is_relative_to(ignore_dir):
            server.watch(path)

    # Start watching `watched_dir` updating mkdocs watches for files that are not in `ignored_dir`
    handler = watchdog.events.FileSystemEventHandler()
    handler.on_moved = callback
    handler.on_deleted = callback
    handler.on_created = callback
    server.observer.schedule(handler, watched_dir, recursive=True)
```
